### ISSUE TITLE
torch.compile fails with dtype mismatch (Half vs Float) in matrix multiplication operations

### ISSUE THREAD

**Issue by Blooming-Tree** [pytorch/pytorch#166647]
### 🐛 Describe the bug

Description
Problem: torch.compile fails during graph compilation when model parameters have mixed data types (float16 and float32), specifically in matrix multiplication operations.
code:
'''
import os
import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np

class BoltzmannMachine(nn.Module):

    def __init__(self, num_visible, num_hidden, learning_rate=0.01, k=1):
        super(BoltzmannMachine, self).__init__()
        self.num_visible = num_visible
        self.num_hidden = num_hidden
        self.k = k
        self.learning_rate = learning_rate
        self.weights = nn.Parameter(torch.randn(num_visible, num_hidden) * 0.01)
        self.visible_bias = nn.Parameter(torch.zeros(num_visible))
        self.hidden_bias = nn.Parameter(torch.zeros(num_hidden, dtype=torch.float16))

    def sample_hidden(self, visible_prob):
        hidden_activations = torch.matmul(visible_prob, self.weights) + self.hidden_bias
        hidden_prob = torch.sigmoid(hidden_activations)
        return (torch.bernoulli(hidden_prob), hidden_prob)

    def sample_visible(self, hidden_prob):
        visible_activations = torch.matmul(hidden_prob, self.weights.t()) + self.visible_bias
        visible_prob = torch.sigmoid(visible_activations)
        return (torch.bernoulli(visible_prob), visible_prob)

    def contrastive_divergence(self, visible_data):
        batch_size = visible_data.shape[0]
        pos_hidden_prob = torch.sigmoid(torch.matmul(visible_data, self.weights) + self.hidden_bias)
        pos_hidden_sample = torch.bernoulli(pos_hidden_prob)
        pos_associations = torch.matmul(visible_data.t(), pos_hidden_prob)
        neg_visible = visible_data.clone()
        for _ in range(self.k):
            (neg_hidden_sample, neg_hidden_prob) = self.sample_hidden(neg_visible)
            (neg_visible_sample, neg_visible_prob) = self.sample_visible(neg_hidden_sample)
            neg_visible = neg_visible_prob
        neg_associations = torch.matmul(neg_visible.t(), neg_hidden_prob)
        self.weights.data += self.learning_rate * (pos_associations - neg_associations) / batch_size
        self.visible_bias.data += self.learning_rate * torch.mean(visible_data - neg_visible, dim=0)
        self.hidden_bias.data += self.learning_rate * torch.mean(pos_hidden_prob - neg_hidden_prob, dim=0)
        error = torch.mean((visible_data - neg_visible) ** 2)
        return error.item()

    def free_energy(self, v):
        vbias_term = torch.matmul(v, self.visible_bias)
        wx_b = torch.matmul(v, self.weights) + self.hidden_bias
        hidden_term = torch.sum(torch.log(1 + torch.exp(wx_b)), dim=1)
        return -hidden_term - vbias_term

    def forward(self, visible_data, n_gibbs=100):
        samples = visible_data.clone()
        for _ in range(n_gibbs):
            (_, hidden_prob) = self.sample_hidden(samples)
            (samples, _) = self.sample_visible(hidden_prob)
        return samples

def get_default_model():
    num_visible = 100
    num_hidden = 50
    learning_rate = 0.01
    k = 1
    model = BoltzmannMachine(num_visible=num_visible, num_hidden=num_hidden, learning_rate=learning_rate, k=k)
    return model

def get_sample_inputs():
    batch_size = 4
    num_visible = 100
    x = torch.randn(batch_size, num_visible)
    return (x,)

def main():
    model = get_default_model()
    model.eval()
    inputs = get_sample_inputs()

    with torch.no_grad():
        output = model(*inputs)
    print('Model executed successfully!')
    print(f'Input shape: {inputs[0].shape}')
    print(f'Output shape: {output.shape}')
    print(f'Model parameters: {sum((p.numel() for p in model.parameters()))}')

    print("\n--- Compile Comparison ---")

    try:
        compiled_model = torch.compile(model)
        compiled_model.eval()
        with torch.no_grad():
            compiled_output = compiled_model(*inputs)

        print('Compiled model: SUCCESS')
        diff = torch.abs(output - compiled_output).max().item()
        print(f'Output difference: {diff:.10f}')

    except Exception as e:
        print(f'Compiled model: FAILED - {e}')

if __name__ == '__main__':
    main()
'''
output:
'''
Model executed successfully!
Input shape: torch.Size([4, 100])
Output shape: torch.Size([4, 100])
Model parameters: 5150

--- Compile Comparison ---
 Compiled model: FAILED - self and mat2 must have the same dtype, but got Half and Float
'''

### Versions

PyTorch version: 2.7.1+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.5.0-6ubuntu2) 9.5.0
Clang version: Could not collect
CMake version: version 4.0.3
Libc version: glibc-2.39

Python version: 3.9.7 (default, Jul 16 2025, 16:34:47) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.14.0-29-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4060 Laptop GPU
Nvidia driver version: 580.65.06
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Address sizes: 39 bits physical, 48 bits virtual
Byte Order: Little Endian
CPU(s): 32
On-line CPU(s) list: 0-31
Vendor ID: GenuineIntel
Model name: Intel(R) Core(TM) i9-14900HX
CPU family: 6
Model: 183
Thread(s) per core: 2
Core(s) per socket: 24
Socket(s): 1
Stepping: 1
CPU(s) scaling MHz: 31%
CPU max MHz: 5800.0000
CPU min MHz: 800.0000
BogoMIPS: 4838.40
Flags: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities
Virtualization: VT-x
L1d cache: 896 KiB (24 instances)
L1i cache: 1.3 MiB (24 instances)
L2 cache: 32 MiB (12 instances)
L3 cache: 36 MiB (1 instance)
NUMA node(s): 1
NUMA node0 CPU(s): 0-31
Vulnerability Gather data sampling: Not affected
Vulnerability Ghostwrite: Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit: Not affected
Vulnerability L1tf: Not affected
Vulnerability Mds: Not affected
Vulnerability Meltdown: Not affected
Vulnerability Mmio stale data: Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed: Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2: Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds: Not affected
Vulnerability Tsx async abort: Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.0
[pip3] nvidia-cublas-cu11==11.10.3.66
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu11==11.7.101
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu11==11.7.99
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu11==11.7.99
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu11==8.5.0.96
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu11==10.9.0.58
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu11==10.2.10.91
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu11==11.4.0.1
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu11==11.7.4.91
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu11==2.14.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu11==11.7.91
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.7.1
[pip3] torchdiffeq==0.2.5
[pip3] torchvision==0.22.1
[pip3] triton==3.3.1
[conda] Could not collect

cc @chauhang @penguinwu
